README
Fedora 18 ARM for Allwinner A10 devices README
----------------------------------------------

Quickstart guide
----------------

1) Insert an sdcard, note any data on the card will be destroyed!
2) Make sure the card is not mounted, run "mount" and if the card shows
   up in the output umount its partitions
3) Write the img file to the card, ie as root do:
   xzcat Fedora-18-Beta-a10-armhfp.img.xz > /dev/mmcblk0
   sync
4) The card is not yet ready for use! Since the A10 u-boot is board
   specific, the image comes without any uboot install, follow the next
   steps to install the right u-boot for your board
5) Remove the card, and re-insert it. The uboot partition should get
   automatically mounted, if not mount it manually,
6) As root (or through sudo) run:
   sh <uboot-part-mount>/select-board.sh
   Note the "sh " in front is necessary because the uboot vfat partition
   often gets auto-mounted noexec. ie:
   sudo sh /run/media/hans/uboot/select-board.sh

   If you've dialog installed the select-board.sh script will prompt for
   your board. If you don't have dialog installed, it will print the list
   of supported boards. Lookup your board and re-run the script with the
   shortname for your board as argument, ie:
   sudo sh /run/media/hans/uboot/select-board.sh mk802
7) umount the uboot and rootfs partitions, ie:
   umount /run/media/hans/uboot
   umount /run/media/hans/rootfs
8) Your sdcard is now ready for use
9) *Before* powering up your A10 device connect it to an hdmi or dvi monitor
10) When first booting from the sdcard inserted Fedora will automatically
   reboot once, this is part of the process to resize the root partition to
   fill the entire sdcard and is normal behavior.
11) After the automatic reboot Fedora will start with the firstboot wizard:
   11a) Configure a user here
   11b) Most A10 devices do not come with a battery backed rtc, so you will
        want to enable ntp. Simply click ok in the ntp dialog. Now firstboot
        will try to start ntp. Since most devices have only wireless, and
        that is not configured yet at this time, click on cancel when it
        tries to start ntp. If it then asks if you want to re-do your ntp
        configuration choose no.
12) Log in as the just created user
13) If you need to configure wireless click on the network icon in the
    top right corner (next to the clock) and select the wireless network
    you want to connect to
14) Enjoy Fedora on your A10 device


Supported Devices:
------------------

Fedora 18 ARM for Allwinner A10 has been tested with the following devices:
* Gooseberry development board
* Mele a1000g/a2000g 1024 MB RAM
* Mini-X 1024 MB RAM
* mk802 (with female mini hdmi) 512 MB RAM
* mk802ii (with male normal hdmi) 1024 MB RAM

Fedora 18 ARM should also work on the following devices:
* Cubieboard development board 512 MB RAM
* Cubieboard development board 1024 MB RAM
* H6 netbook
* Hackberry development board
* Hyundai a7hd tablet
* Mele a1000/a2000 512 MB RAM
* Mini-X 512 MB RAM


Configuring the display output
------------------------------

Multiple video outputs at the same time are not supported. By default
hdmi output with EDID is used for all devices, except for tablets/netbooks
where the default output is the lcd.

The default hdmi output with EDID will get the native resolution of your
TV / monitor and use that. Note that in order for this to work your TV /
monitor must be connected *and turned on*, before booting your device.

The output resolution can be configured with the disp.screen0_output_mode
kernel cmdline value, which can be found in the extrargs part of uEnv.txt in
the uboot partition. The default uEnv.txt contains the following value:
disp.screen0_output_mode=EDID:1280x720p60

This means try to use EDID and if no valid EDID info is found fallback to
1280x720p60.

The used output can be changed by adding disp.screen0_output_type=X to the
extraargs in uEnv.txt. With X being one of: 0:none; 1:lcd; 2:tv; 3:hdmi; 4:vga

Some per display type notes:
-lcd outputs: Hardcoded to the native mode, disp.screen0_output_mode is ignored
-tv: For the cvbs output disp.screen0_output_mode must be set to one of the
 following: pal, pal-svideo, ntsc, ntsc-svideo, pal-m, pal-m-svideo, pal-nc,
 pal-nc-svideo. Note the -svideo variants should only be used on boards with
 an svideo connector, for composite out use the regular variants, ie:
 disp.screen0_output_type=2 disp.screen0_output_mode=pal
-hdmi: To override the EDID detected mode, drop the "EDID:" from the
 disp.screen0_output_mode value and set it to the desired mode, ie:
 disp.screen0_output_type=3 disp.screen0_output_mode=1360x768p60
-vga: Does not support EDID, "EDID:" must be removed from the
 disp.screen0_output_mode value otherwise it will be ignored. interlaced 
 progressive and refreshrate settings specified are ignored, each resolution
 has hardcoded values for these. Example usage:
 disp.screen0_output_type=4 disp.screen0_output_mode=1024x768


USB controller caveats
----------------------
The OTG USB controller in host mode only supports a limited number of
devices, plugging in a hub + mouse + keyboard typically will make either
the mouse or keyboard not work. This is a hardware limitation which we
will likely not be able to work around.

On tv-sticks and top-set boxes, simply avoid the otg connector, instead
use a hub in a regular host usb connector. Note on the mini-x the otg / host
marking is not always correct. If things don't work try using the OTG
connector instead!

On tablets and the gooseberry unfortunately only the otg connector is
available. One solution there is using a single usb-device which is
both a keyboard and a mouse at the same time. IE the receiver for logitech
wireless desktop sets.


Supported hardware components / features:
-----------------------------------------

Fedora 18 ARM for Allwinner A10 supports the following components:
* CPU + PMU + RAM
* Serial ports
* MMC cards
* Internal NAND storage
* Framebuffer on lcd / vga / hdmi / composite video
* Sound both analog out and over hdmi
* OTG USB controller
* Both standard USB host controllers
* Wifi
* Wired Ethernet
* SATA
* IR (untested at this time)


Unsupported hardware components:
--------------------------------

The following components require various proprietary blobs to be used, and
as such are not supported in the Fedora images. The kernel drivers for them
are present (usually as modules), so if you add the necessary blobs you
might get these to work:
* Mali 400 GPU
* Cedar hardware video & audio decoding and encoding engine
* G2D 2d engine


Differences from stock Fedora
-----------------------------
* Since the A10 is not a very powerful soc some services which are enabled by
  default on Fedora are disabled in the image, see build-image.sh for a list.
* No plymouth: we log to a serial console for debugging so no pretty splash.
  Also we don't use an initrd, so removing the console=ttyS0,115200 from
  the extraargs in uEnv.txt will give plymouth, but so late it hardly matters.


Future plans / todo:
--------------------
* Currently 192MiB of the A10's RAM is reserved, leaving only 320MiB for Linux
  use on 512MiB RAM systems. This is mostly caused by memory reserved for use
  by hardware listed under the "Unsupported hardware components" paragraph.
  Since we're not using those blocks anyways the plan is to give this RAM back
  to Linux by adding kernel cmdline options to reserve the RAM, and then:
  1) On Fedora by default set all the reserved sizes to 0 on the cmdline
  2) The kernel modules for these hw-blocks will check the reserved size and
     if it is 0 refuse to load
  3) Users who want to use these hw-blocks already need to manually configure
     the necessary blobs, they will now also need to adjust the kernel cmdline
* Currently both the analog and hdmi audio outputs of the A10 are both
  always registered. This means that the user must always manually pick the
  right soundcard for his config. The plan is to instead only register the
  analog out on boards which actually have an analog out connecter, and only
  register the hdmi audio card when an hdmi-audio capable monitor is detected


Rebuilding the Fedora 18 ARM for Allwinner A10 disk image
---------------------------------------------------------

Building the Fedora 18 ARM for Allwinner A10 disk image consists of 2 steps
1) Building a uboot.tar.gz and rootfs.tar.gz "overlays", this is done
   bu the build-boot-root-sh script
2) Combining uboot.tar.gz and rootfs.tar.gz with an official Fedora 18
   ARM panda image. The a10 image you downloaded is named the same as the panda
   image used as base with panda replaced by a10, so to get the panda image
   name replace a10 with panda again :) This combining is done by the
   build-image.sh script

These scripts are hosted here:
https://github.com/jwrdegoede/sunxi-fedora-scripts.git

A copy of the exact versions of these scripts used to build this Fedora A10
image can be found in the scripts directory of the uboot partition, the
kernel config used during the build can be found here too.

If you want to exactly reproduce this image it is important to use the
scripts from the scripts dir of the uboot partition, as the scripts contain
GIT tags used during the build to checkout the exact versions to build.

The pre-conditions these scripts expect to be met, and the exact usage of
them is documented in comments in the top of each script.


ChangeLog
Fedora-18-Beta-a10-armhfp-r2 (15 January 2013):
-----------------------------------------------
-Fix rootfs customization script not being run during image generation:
 -Use sunxi instead of panda in hostname 
 -Disable a bunch of services as the sunxi is not all that fast
-Update yum.conf to exclude kernel-omap* updates, so as to not replace the
 sunxi kernel with an omap kernel
-Various kernel improvements:
 -Fix video modes with certain pixel-clocks not working under X, ie
  1600x1200 with a 130000000 Hz pixelclock
 -Fix tv-out not working

Fedora-18-Beta-a10-armhfp (15 January 2013):
--------------------------------------------
-Initial release based on Fedora-18-Beta-panda-armhfp.img


build-image.sh
#!/bin/sh
#
# This script builds a Fedora ARM Allwinner sunxi image, as input it takes
# uboot.tar.gz and rootfs.tar.gz tarbals as build by build-boot-root.sh
# and an *unzipped* Fedora ARM panda sdcard image
#
# This script expects to be run from a directory which contains the
# uboot.tar.gz and rootfs.tar.gz files
#
# This script expects /dev/$LOOP to be free for its use

set -e

if [ "$1" = "--nocopy" ]; then
   NOCOPY=1
   shift
fi

if [ $# != 2 ]; then
   echo "Usage $0 <Fedora-arm-panda-img-in> <Fedora-arm-a10-img-out>"
   exit 1
fi

if [ "$(id -u)" != 0 ]; then
   echo "$0 must be run as root to be able to setup loop devices"
   exit 1
fi

LOOP=loop5
IMG_IN="$1"
IMG_OUT="$2"
UBOOT="$(pwd)/uboot"
ROOTFS="$(pwd)/rootfs"

if [ -z "$NOCOPY" ]; then
    echo "Copying $IMG_IN to $IMG_OUT"
    cp "$IMG_IN" "$IMG_OUT"
fi

echo "Setting up loopback mount of $IMG_OUT and its partitions"
losetup "/dev/$LOOP" "$IMG_OUT"
kpartx -a "/dev/$LOOP"
udevadm settle
mkdir -p "$UBOOT"
mount "/dev/mapper/${LOOP}p1" "$UBOOT"
mkdir -p "$ROOTFS"
mount "/dev/mapper/${LOOP}p3" "$ROOTFS"

echo "Clearing uboot area"
dd if=/dev/zero of="/dev/$LOOP" bs=1024 seek=8 count=1016

echo "Cleaning panda uboot files and copying sunxi uboot files"
pushd "$UBOOT" > /dev/null
rm -rf *
tar xfz "$UBOOT.tar.gz" --no-same-owner --no-same-permissions
popd > /dev/null

echo "Cleaning panda specific rootfs files and copying sunxi rootfs files"
pushd "$ROOTFS" > /dev/null
rm -rf usr/lib/modules/*
tar xfz "$ROOTFS.tar.gz" --no-overwrite-dir
echo "Customizing sunxi rootfs"
sed -i 's/panda/sunxi/' etc/hostname etc/sysconfig/network
sed -i 's/installonly_limit=3/installonly_limit=3\nexclude=kernel-omap*/' etc/yum.conf
for i in 'abrt*' atd irqbalance mdmonitor rpcbind sendmail sm-client smartd; do
    rm etc/systemd/system/multi-user.target.wants/$i.service
done
rm etc/systemd/system/spice-vdagentd.target.wants/spice-vdagentd.service
rm etc/systemd/system/sysinit.target.wants/mdmonitor-takeover.service
popd > /dev/null

echo "Cleaning up loopback mounts"
umount "$ROOTFS"
umount "$UBOOT"
kpartx -d "/dev/$LOOP"
losetup -d "/dev/$LOOP"

echo "$IMG_OUT is ready for dd-ing to an sdcard"


build-boot-root.sh
#!/bin/sh
#
# This scripts builds an Allwinner A10 kernel + per board u-boot, spl and fex
# It will then place all the build files into 2 directories:
# $DESTDIR/uboot and $DESTDIR/rootfs
# and then tars up these directories to:
# $DESTDIR/uboot.tar.gz and $DESTDIR/rootfs.tar.gz
# Note that it also leaves the directories in place for easy inspection
#
# These tarbals are intended to be untarred to respectively the uboot and
# rootfs partition of a Fedora panda sdcard image, thereby turning this image
# into an Fedora A10 sdcard image. See build-image.sh for a script automating
# this.
#
# The latest version of this script can be found here:
# https://github.com/jwrdegoede/sunxi-fedora-scripts.git
#
# To get the exact same versions as used on your sdcard, use the copy of
# this script found on your sdcard, as that contains all the git-tags used
# to build the sdcard image.
#
# This script must be run under Fedora-18 x86_64, with the following
# packages installed:
# gcc-arm-linux-gnu
# uboot-tools
#
# Also the fex2bin utility from:
# https://github.com/linux-sunxi/sunxi-tools.git
# (not yet packaged) needs to be available in the PATH somewhere
#
# This script must be run from a directory which contains clones of the
# following git repositories:
# https://github.com/jwrdegoede/sunxi-fedora-scripts.git
# https://github.com/jwrdegoede/u-boot-sunxi.git
# https://github.com/jwrdegoede/sunxi-boards.git
# https://github.com/jwrdegoede/sunxi-kernel-config.git
# https://github.com/jwrdegoede/linux-sunxi.git

KERNER_VER=3.4
A10_BOARDS="cubieboard cubieboard_512 gooseberry_a721 h6 hackberry hyundai_a7hd mele_a1000 mele_a1000g mini-x mini-x-1gb mk802 mk802ii"
UBOOT_TAG=fedora-18-13012013
KERNEL_CONFIG_TAG=fedora-18-17012013
KERNEL_TAG=fedora-18-17012013
A10_BOARDS_TAG=fedora-18-13012013-3
SCRIPTS_TAG=fedora-18-17012013

if [ -z "$DESTDIR" ]; then
    DESTDIR=$(pwd)
fi

set -e
set -x

[ -d $DESTDIR/uboot ] && rm -r $DESTDIR/uboot
[ -d $DESTDIR/rootfs ] && rm -r $DESTDIR/rootfs
mkdir $DESTDIR/uboot
mkdir $DESTDIR/rootfs

pushd u-boot-sunxi
git checkout $UBOOT_TAG
[ "$1" != --noclean ] && git clean -dxf
mkdir $DESTDIR/uboot/boards
# Note the changing board configs always force a rebuild
for i in $A10_BOARDS; do
    make -j4 CROSS_COMPILE=arm-linux-gnu- O=$i distclean
    make -j4 CROSS_COMPILE=arm-linux-gnu- O=$i $i
    mkdir $DESTDIR/uboot/boards/$i
    cp $i/spl/sunxi-spl.bin $DESTDIR/uboot/boards/$i
    cp $i/u-boot.bin $DESTDIR/uboot/boards/$i
done
popd

pushd sunxi-boards
git checkout $A10_BOARDS_TAG
[ "$1" != --noclean ] && git clean -dxf
for i in $A10_BOARDS; do
    cp -p sys_config/a10/$i.fex $DESTDIR/uboot/boards/$i
    fex2bin sys_config/a10/$i.fex $DESTDIR/uboot/boards/$i/script.bin
done
popd

pushd sunxi-kernel-config
git checkout $KERNEL_CONFIG_TAG
[ "$1" != --noclean ] && git clean -dxf
make VERSION=$KERNER_VER -f Makefile.config kernel-$KERNER_VER-armv7hl-a10.config
popd

pushd linux-sunxi
git checkout $KERNEL_TAG
[ "$1" != --noclean ] && git clean -dxf
cp -a ../sunxi-kernel-config/kernel-$KERNER_VER-armv7hl-a10.config .config
make ARCH=arm CROSS_COMPILE=arm-linux-gnu- CONFIG_DEBUG_SECTION_MISMATCH=y -j4 uImage modules
mkdir $DESTDIR/rootfs/usr
make ARCH=arm CROSS_COMPILE=arm-linux-gnu- INSTALL_MOD_PATH=$DESTDIR/rootfs/usr modules_install
for i in `find $DESTDIR/rootfs/usr/lib/modules -name "*.ko"`; do
    arm-linux-gnu-strip --strip-debug "$i"
done
mkdir $DESTDIR/uboot/scripts
cp arch/arm/boot/uImage $DESTDIR/uboot
cp .config $DESTDIR/uboot/scripts/kernel-$KERNER_VER-armv7hl-a10.config
popd

pushd sunxi-fedora-scripts
git checkout $SCRIPTS_TAG
[ "$1" != --noclean ] && git clean -dxf
../u-boot-sunxi/mele_a1000/tools/mkenvimage -s 131072 \
  -o $DESTDIR/uboot/boards/uEnv-img.bin uEnv-full.txt
mkimage -C none -A arm -T script -d boot.cmd $DESTDIR/uboot/boot.scr
cp -p boot.cmd README select-board.sh $DESTDIR/uboot
cp -p uEnv-boot.txt $DESTDIR/uboot/uEnv.txt
cp -p build-boot-root.sh build-image.sh $DESTDIR/uboot/scripts
# replace rootfs-resize with one which understands running without an initrd
mkdir $DESTDIR/rootfs/usr/sbin
cp -p rootfs-resize $DESTDIR/rootfs/usr/sbin
popd

echo
echo "Successfully build uboot and rootfs directories, packing ..."

pushd $DESTDIR/uboot
tar cfz $DESTDIR/uboot.tar.gz *
popd

pushd $DESTDIR/rootfs
tar cfz $DESTDIR/rootfs.tar.gz *
popd

echo "Successfully generated uboot.tar.gz and rootfs.tar.gz"


select-board.sh
#!/bin/sh

DIALOG="$(which dialog 2> /dev/null)"

set -e

CANON="$(readlink -f $0)"
BOARD="$1"
UBOOT_MOUNT="$(dirname $CANON)"
BOARDS_DIR="$UBOOT_MOUNT/boards"
UBOOT_DEV="$(df $CANON | tail -n 1 | awk '{print $1}')"

BOARDS=()
BOARDS+=(cubieboard_512  "Cubieboard development board 512 MB RAM")
BOARDS+=(cubieboard      "Cubieboard development board 1024 MB RAM")
BOARDS+=(gooseberry_a721 "Gooseberry development board")
BOARDS+=(h6              "H6 netbook")
BOARDS+=(hackberry       "Hackberry development board")
BOARDS+=(hyundai_a7hd    "Hyundai a7hd tablet")
BOARDS+=(mele_a1000      "Mele a1000/a2000 512 MB RAM")
BOARDS+=(mele_a1000g     "Mele a1000g/a2000g 1024 MB RAM")
BOARDS+=(mini-x          "Mini-X 512 MB RAM")
BOARDS+=(mini-x-1gb      "Mini-X 1024 MB RAM")
BOARDS+=(mk802           "mk802 (with female mini hdmi) 512 MB RAM")
BOARDS+=(mk802ii         "mk802ii (with male normal hdmi) 1024 MB RAM")

if [ "$1" = "--help" -o -z "$DIALOG" -a -z "$BOARD" ]; then
    echo "Usage: \"$0 <board>\""
    echo "Available boards:"
    for (( i = 0; i < ${#BOARDS[@]} ; i+=2 )); do
        printf "%-20s%s\n" "${BOARDS[$i]}" "${BOARDS[(($i + 1))]}"
    done
    exit 0
fi


# Remove partition at the end of UBOOT_DEV to get the sdcard dev
if expr match "$UBOOT_DEV" "/dev/sd.1" > /dev/null; then
    SDCARD_DEV=$(echo $UBOOT_DEV | sed 's+1$++')
elif expr match "$UBOOT_DEV" "/dev/mmcblk.p1" > /dev/null; then
    SDCARD_DEV=$(echo $UBOOT_DEV | sed 's+p1$++')
else
    echo "Error cannot determine sdcard-dev from uboot-dev $UBOOT_DEV"
    exit 1
fi

if [ ! -w "$SDCARD_DEV" ]; then
    echo "Error cannot write to $SDCARD_DEV (try running as root)"
    exit 1
fi


yesno () {
    if [ -z "$DIALOG" ]; then
        echo "$1"
        echo -n "Press enter to continue, CTRL+C to cancel"
        read
    else
        dialog --yesno "$1" 20 70
    fi
}


if [ -z "$BOARD" ]; then
    yesno "Your sdcard has been detected at $SDCARD_DEV. If this is wrong this utility may corrupt data on the detected disk! Is $SDCARD_DEV the correct disk to install the spl, u-boot and uEnv too ?"

    TMPFILE=$(mktemp)
    dialog --menu "Select your Allwinner board" 20 70 30 "${BOARDS[@]}" 2> $TMPFILE
    BOARD="$(cat $TMPFILE)"
    rm $TMPFILE
fi

if [ ! -d $BOARDS_DIR/$BOARD ]; then
    echo "Error cannot find board dir: $BOARDS_DIR/$BOARD"
    exit 1
fi

yesno "Are you sure you want to install the spl, u-boot and uEnv for $BOARD from $BOARDS_DIR onto $SDCARD_DEV ?"

echo
echo "Installing spl, u-boot and uEnv for $BOARD onto $SDCARD_DEV ..."

dd if="$BOARDS_DIR/$BOARD/sunxi-spl.bin" of="$SDCARD_DEV" bs=1024 seek=8
dd if="$BOARDS_DIR/$BOARD/u-boot.bin" of="$SDCARD_DEV" bs=1024 seek=32
dd if="$BOARDS_DIR/uEnv-img.bin" of="$SDCARD_DEV" bs=1024 seek=544
cp "$BOARDS_DIR/$BOARD/script.bin" "$UBOOT_MOUNT"
sync

echo "Done"
